### PR TITLE
First draft of an analyzer for capability APIs

### PR DESCRIPTION
**Pull request by terrajobst**
This design isn't ready yet. I only created the draft PR so that I can point folks to it.

### CODE REVIEW

File: accepted/capability-checks/capability-checks.md [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+have to suppress them. One option is add typical capabilities to the assembly,
+but that's messy. Another option is tag capabilities with platform annotations
+that basically say "on this platform, you don't have to ask for capability
+checks". Then it's only a matter of flowing the platform information from the
+build into the analyzer.
+
+#### Should this analyzer be-opt in?
+
+This might be useful for .NET Standard library developers and much less useful
+for people building .NET Core apps. We might also want different defaults.
```
Comment (vitek-karas) [inline accepted/capability-checks/capability-checks.md:44]: Some capabilities will be less interesting for .NET Core apps, but some are as applicable as for .NET Standard libraries (I can still try to use Registry APIs when building app for Linux, it should warn). For that reason I think this should be on-by-default if we can make it performant enough.

File: accepted/capability-checks/capability-checks.md [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+#### Should we provide an assert-style attribute?
+
+This would behave the same was tagging the assemblies, modules, types, or
+members with the `[RequiredXxxCapability]` but without requiring callers to
+check it. It's basically a suppression for a specific capability.
```
Comment (vitek-karas) [inline accepted/capability-checks/capability-checks.md:50]: If it looks like suppression and behaves like suppression -> it might as well be a suppression 😄  - Why would not `SuppressMessage` (or the new `UnconditionalSuppressMessage`) work here? Maybe I don't understand the "assertion" pattern.

File: accepted/capability-checks/capability-checks.md [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+#### Should different capabilities have different diagnostic ids?
+
+This probably doesn't work in Roslyn as the diagnostic set needs to be known
+statically.
+
```
Comment (vitek-karas) [inline accepted/capability-checks/capability-checks.md:56]: In general I agree, but it would be nice if we could make at least some with distinct IDs. If nothing else it would behave nicer in combination with suppression and warning settings in IDEs.
So for example those we define in the framework could come with distinct IDs.

File: accepted/capability-checks/capability-checks.md [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        var value = key.GetValue("foo");
+
+        /*
+        using (var key = Registry.CurrentUser.OpenSubKey(@"SoftwareFabrikamAssetManagement"))
+        {
+            if (key?.GetValue("LoggingDirectoryPath") is string configuredPath)
+                return configuredPath;
+        }
+        */s
+    }
```
Comment (vitek-karas) [inline accepted/capability-checks/capability-checks.md:100]: What if the code looks like:
```C#
if (Registry.IsSupported)
{ // read from registry }
else
{ throw new Exception("This feature requires registry"); }
```

The analyzer would probably mark this as "perfectly fine", but in reality it's as broken as before. Should we consider some kind of detection of this "Anti-pattern"?

File: accepted/capability-checks/capability-checks.md [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+    // Would also mark
+    // - Other *Builder types
+    // - AppDomain.DefineDynamicAssembly
+    // - ILGenerator
+}
+```
+
+## Q & A
```
Comment (vitek-karas) [inline accepted/capability-checks/capability-checks.md:234]: Should we possibly fold the proposed "linker unfriendly" attribute into this as well?: https://github.com/dotnet/runtime/issues/33862
It looks like a good fit.

Comment (vitek-karas) [inline accepted/capability-checks/capability-checks.md:234]: What is the handling of issues inside dependencies (NuGets mostly)? For example with the registry APIs above:
I write an app which itself doesn't call any of the registry APIs, but it uses a NuGet which does call them. Running the analyzer on the app itself will not reveal any problems, but the app is still broken on Linux.

Comment (vitek-karas) [inline accepted/capability-checks/capability-checks.md:234]: We should also consider how is this related to the feature switch functionality: https://github.com/dotnet/designs/blob/master/accepted/2020/feature-switch.md

In short - linker could remove the "false" branches in lot of cases (reducing code size) - assuming this is integrated correctly.

**Comment by terrajobst** [comment]
It doesn't seem we're pursuing this for now. If scenarios requires this, we might come back to it.
